ext/LinearSolveHYPRE.jl

```diff
@@ -0,0 +1,193 @@
+module LinearSolveHYPRE
+
+using HYPRE.LibHYPRE: HYPRE_Complex
+using HYPRE: HYPRE, HYPREMatrix, HYPRESolver, HYPREVector
+using IterativeSolvers: Identity
+using LinearSolve: HYPREAlgorithm, LinearCache, LinearProblem, LinearSolve,
+                   OperatorAssumptions, default_tol, init_cacheval, issquare, set_cacheval
+using SciMLBase: LinearProblem, SciMLBase
+using UnPack: @unpack
+
+mutable struct HYPRECache
+    solver::Union{HYPRE.HYPRESolver, Nothing}
+    A::Union{HYPREMatrix, Nothing}
+    b::Union{HYPREVector, Nothing}
+    oidb::UInt # objectid(b) to keep track of isfresh for b
+    u::Union{HYPREVector, Nothing}
+end
+
+function LinearSolve.init_cacheval(alg::HYPREAlgorithm, A, b, u, Pl, Pr, maxiters::Int,
+                                   abstol, reltol,
+                                   verbose::Bool, assumptions::OperatorAssumptions)
+    return HYPRECache(nothing, nothing, nothing, UInt(0), nothing)
+end
+
+# Note:
+# SciMLBase.init is overloaded here instead of just LinearSolve.init_cacheval for two
+# reasons:
+# - HYPREArrays can't really be `deepcopy`d, so that is turned off by default
```

Comment:
Why are they copied unconditionally? And why is this copy using `deepcopy` and not `copy`? IIUC `deepcopy` shouldn't be used for things like this.

It seems more useful to let the algorithm decide, many solvers don't modify the matrix, so it is a waste to copy it, no??